server/src/authentication/middleware/authentication.js

```diff
@@ -242,141 +242,158 @@ module.exports = function (passport) {
+    );
+  } else {
+    passport.use(
+      new OIDCStrategy(
+        {
+          identityMetadata: config.authentication.microsoft.identityMetadata,
+          clientID: config.authentication.microsoft.clientID,
+          responseType: config.authentication.microsoft.responseType,
+          responseMode: config.authentication.microsoft.responseMode,
+          redirectUrl: config.authentication.microsoft.redirectUrl,
+          allowHttpForRedirectUrl:
+            config.authentication.microsoft.allowHttpForRedirectUrl,
+          clientSecret: config.authentication.microsoft.clientSecret,
+          validateIssuer: config.authentication.microsoft.validateIssuer,
+          isB2C: config.authentication.microsoft.isB2C,
+          issuer: config.authentication.microsoft.issuer,
+          passReqToCallback: config.authentication.microsoft.passReqToCallback,
+          scope: config.authentication.microsoft.scope,
+          loggingLevel: config.authentication.microsoft.loggingLevel,
+          nonceLifetime: config.authentication.microsoft.nonceLifetime,
+          nonceMaxAmount: config.authentication.microsoft.nonceMaxAmount,
+          useCookieInsteadOfSession:
+            config.authentication.microsoft.useCookieInsteadOfSession,
+          cookieEncryptionKeys:
+            config.authentication.microsoft.cookieEncryptionKeys,
+          clockSkew: config.authentication.microsoft.clockSkew,
+          loggingNoPII: false,
+        },
+        async (iss, sub, profile, accessToken, refreshToken, done) => {
+          const provider = "azure-ad";
+          const oid = profile._json.oid;
+          const email = profile._json.email;
+          let user = await UserModel.findOne({ provider, providerId: oid });
+          if (user) {
+            log.info("Azure-AD user found");
+            return done(null, user);
+          }
+          user = await UserModel.findOne({ email });
+          if (user) {
+            if (user.state === UserStates.VERIFY) {
+              log.warn("Waiting for email verification");
+              return done(null, false, {
+                message: texts.allreadyWaitingForEmailVerification,
+              });
+            } else {
+              log.warn("Known email");
+              return done(null, false, {
+                message: formatString(texts.wrongProvider, user.provider),
+              });
+            }
           }
+          log.info("Updating Azure-AD user");
+          log.info("Create Azure-AD user");
+          user = await UserModel.create({
+            oid,
+            state: UserStates.ACTIVE,
+            roles: DefaultRoles,
+            provider,
+            providerId: oid,
+            email,
+          });
+          await emails.sendWelcomeEmail(user.email);
+          done(null, user);
         }
+      )
+    );
+  }
+
+  if (
+    !config.authentication.microsoft.clientID ||

```

Comment:
Should this be `config.authentication.google.clientIID` instead of `config.authentication.microsoft.clientID`?

Reply:
Yes indeed